Add Storage::HmacKey

quartzmo opened the pull request:
WIP. Do not merge.

[closes #3281, closes #3442]

quartzmo commented on google-cloud-storage/lib/google/cloud/storage/hmac_key.rb:
```diff
@@ -0,0 +1,309 @@
+        # @param [Google::Apis::StorageV1::HmacKeyMetadata] gapi
+        #
+        #
+        def self.from_gapi_metadata gapi, service, user_project: nil
```
@blowmage Are you OK with this second "mode" when the instance is initialized only with the `Google::Apis::StorageV1::HmacKeyMetadata` data? This is how it is presented by every operation except create, which uses the full representation (`from_gapi`, above).

quartzmo replied:
Only one field is missing in this mode, the HMAC `secret`.

blowmage replied:
As long as it's private and we can refactor it without breaking the public API, so I have no objections.

blowmage commented on google-cloud-storage/lib/google/cloud/storage/hmac_key.rb:
```diff
@@ -164,7 +164,8 @@ def active?
         #
-        # @return [Boolean] Returns `true` if the HMAC key was updated.
+        # @return [Google::Cloud::Storage::HmacKey] Returns the HMAC key for
+        #   chaining.
```
What do you think about changing this to "chaining methods" or "method chaining"?

quartzmo replied:
Updated.

blowmage commented on google-cloud-storage/acceptance/storage/project_test.rb:
```diff
@@ -22,4 +22,80 @@
     # https://stackoverflow.com/questions/22993545/ruby-email-validation-with-regex
     email.must_match /\A[\w+\-.]+@[a-z\d\-]+(\.[a-z\d\-]+)*\.[a-z]+\z/i
   end
+  focus
```
focus

quartzmo replied:
WIP

blowmage commented on google-cloud-storage/acceptance/storage_helper.rb:
```diff
@@ -89,16 +89,16 @@ def try_with_backoff msg = nil, limit: 10
-      end
-      reporter.record result
-    end
+    # def self.run_one_method klass, method_name, reporter
```
Probably want to uncomment this before merging.

quartzmo replied:
WIP

quartzmo commented:
@jkwlui Does this PR look good to you?

quartzmo commented:
I updated `signer_v4_test.rb` test to use [v4_signatures.json](https://github.com/googleapis/conformance-tests/blob/master/storage/v1/v4_signatures.json) from the googleapis/conformance-tests repo, which resulted in 1 broken test:

```
  1) Failure:
SignerV4Test#test_: Headers should be trimmed [/Users/quartzmo/code/google/codez/google-cloud-ruby/google-cloud-storage/test/google/cloud/storage/file/signer_v4_test.rb:39]:
--- expected
+++ actual
@@ -1 +1 @@
-"https://storage.googleapis.com/test-bucket/test-object?X-Goog-Algorithm=GOOG4-RSA-SHA256&X-Goog-Credential=test-iam-credentials%40dummy-project-id.iam.gserviceaccount.com%2F20190201%2Fauto%2Fstorage%2Fgoog4_request&X-Goog-Date=20190201T090000Z&X-Goog-Expires=10&X-Goog-SignedHeaders=collapsed%3Bhost%3Bleading%3Btabs%3Btrailing&X-Goog-Signature=52fa1d70bcd527ee9c1241f87c56bc481526e8a63d440948595ff776faacb0caa6e8a3060b113546cb27ed29d80c88d402947d83948758d4e5c49e47d9482751d46b2a99c2dae5bc8f7baffab03dec05b28b5d605610686c48e867d6a4239a2a61a785df7d6099d155bba57d0d331d66d667b5df8e165e8277e2675678fc28499abd34053a2bc4e4fa21d032c4278fd29897e8307f142506a3d8d07149cded15f7defa77028fb88ff45132cee5f6232feb8e7f899fe361f1f8ceed0795aff860084f35e27475447dc6e64e4baa09e96a725eee6fa3c408d6bb51c2bd5f649afb8339f46997d9ef22496a79cf0846e52ac941c08dc4b9d63639d0ff2ce8637412"
+"https://storage.googleapis.com/test-bucket/test-object?X-Goog-Algorithm=GOOG4-RSA-SHA256&X-Goog-Credential=test-iam-credentials%40dummy-project-id.iam.gserviceaccount.com%2F20190201%2Fauto%2Fstorage%2Fgoog4_request&X-Goog-Date=20190201T090000Z&X-Goog-Expires=10&X-Goog-SignedHeaders=collapsed%3Bhost%3Bleading%3Btabs%3Btrailing&X-Goog-Signature=75d77a3ed2d9b74ff7e1e23b2fd7cc714ad4cc32518c65f3a8197827cd87d302623bab990cf2ff3a633bfaae69b6c2d897add78c105aa68411229610421c4239579add4aff6bdbd5067a0fd61c3aa0029d7de0f8ae88fa3458fa70f875e841d6df9598597d9012b9f848c6857e08f2704ca2f332c71738490ffdda2ed928f9340549d7295745725062d28dc1696eab7cb3b88ac4fd445e951423f645d680a60dd8033d65b65f4c10286f59f4258dbb2bcf36a76ffdd40574104cbbf0b76901c24df5854f24c42e9192fcedc386d85704fec6a6bad3a5201e1fb6c491a4c43371b0913420743580daf3504e99204c6ec894b4d70cd27bc60c3fe2850e8bf3ed22"
```

quartzmo commented:
@jskeet @jdpedrie I updated `signer_v4_test.rb` test to use [v4_signatures.json](https://github.com/googleapis/conformance-tests/blob/master/storage/v1/v4_signatures.json) from the googleapis/conformance-tests repo, which resulted in 1 broken conformance test. Do you see anything wrong with my current (failing) canonical request for the "Headers should be trimmed" test?

```
canonical_request:

GET
/test-bucket/test-object
X-Goog-Algorithm=GOOG4-RSA-SHA256&X-Goog-Credential=test-iam-credentials%40dummy-project-id.iam.gserviceaccount.com%2F20190201%2Fauto%2Fstorage%2Fgoog4_request&X-Goog-Date=20190201T090000Z&X-Goog-Expires=10&X-Goog-SignedHeaders=collapsed%3Bhost%3Bleading%3Btabs%3Btrailing
collapsed:abc def
host:storage.googleapis.com
leading:xyz
tabs:abc def
trailing:abc

collapsed;host;leading;tabs;trailing
UNSIGNED-PAYLOAD
```


```
  1) Failure:
SignerV4Test#test_: Headers should be trimmed [/Users/quartzmo/code/google/codez/google-cloud-ruby/google-cloud-storage/test/google/cloud/storage/file/signer_v4_test.rb:39]:
--- expected
+++ actual
@@ -1 +1 @@
-"https://storage.googleapis.com/test-bucket/test-object?X-Goog-Algorithm=GOOG4-RSA-SHA256&X-Goog-Credential=test-iam-credentials%40dummy-project-id.iam.gserviceaccount.com%2F20190201%2Fauto%2Fstorage%2Fgoog4_request&X-Goog-Date=20190201T090000Z&X-Goog-Expires=10&X-Goog-SignedHeaders=collapsed%3Bhost%3Bleading%3Btabs%3Btrailing&X-Goog-Signature=52fa1d70bcd527ee9c1241f87c56bc481526e8a63d440948595ff776faacb0caa6e8a3060b113546cb27ed29d80c88d402947d83948758d4e5c49e47d9482751d46b2a99c2dae5bc8f7baffab03dec05b28b5d605610686c48e867d6a4239a2a61a785df7d6099d155bba57d0d331d66d667b5df8e165e8277e2675678fc28499abd34053a2bc4e4fa21d032c4278fd29897e8307f142506a3d8d07149cded15f7defa77028fb88ff45132cee5f6232feb8e7f899fe361f1f8ceed0795aff860084f35e27475447dc6e64e4baa09e96a725eee6fa3c408d6bb51c2bd5f649afb8339f46997d9ef22496a79cf0846e52ac941c08dc4b9d63639d0ff2ce8637412"
+"https://storage.googleapis.com/test-bucket/test-object?X-Goog-Algorithm=GOOG4-RSA-SHA256&X-Goog-Credential=test-iam-credentials%40dummy-project-id.iam.gserviceaccount.com%2F20190201%2Fauto%2Fstorage%2Fgoog4_request&X-Goog-Date=20190201T090000Z&X-Goog-Expires=10&X-Goog-SignedHeaders=collapsed%3Bhost%3Bleading%3Btabs%3Btrailing&X-Goog-Signature=75d77a3ed2d9b74ff7e1e23b2fd7cc714ad4cc32518c65f3a8197827cd87d302623bab990cf2ff3a633bfaae69b6c2d897add78c105aa68411229610421c4239579add4aff6bdbd5067a0fd61c3aa0029d7de0f8ae88fa3458fa70f875e841d6df9598597d9012b9f848c6857e08f2704ca2f332c71738490ffdda2ed928f9340549d7295745725062d28dc1696eab7cb3b88ac4fd445e951423f645d680a60dd8033d65b65f4c10286f59f4258dbb2bcf36a76ffdd40574104cbbf0b76901c24df5854f24c42e9192fcedc386d85704fec6a6bad3a5201e1fb6c491a4c43371b0913420743580daf3504e99204c6ec894b4d70cd27bc60c3fe2850e8bf3ed22"
```

jskeet commented:
I'll log the canonical request for the (passing) C# test and let you know.

jskeet commented:
Here's my canonical request, which looks the same as yours as far as I can see...

```text
GET
/test-bucket/test-object
X-Goog-Algorithm=GOOG4-RSA-SHA256&X-Goog-Credential=test-iam-credentials%40dummy-project-id.iam.gserviceaccount.com%2F20190201%2Fauto%2Fstorage%2Fgoog4_request&X-Goog-Date=20190201T090000Z&X-Goog-Expires=10&X-Goog-SignedHeaders=collapsed%3Bhost%3Bleading%3Btabs%3Btrailing
collapsed:abc def
host:storage.googleapis.com
leading:xyz
tabs:abc def
trailing:abc

collapsed;host;leading;tabs;trailing
UNSIGNED-PAYLOAD
```
Note that the "tabs:abc def" should have a tab character between "abc" and "def". Might that be the problem?

quartzmo commented:
@jskeet Thanks!

> the "tabs:abc def" should have a tab character between "abc" and "def". Might that be the problem?

It seems likely to be the issue, since the [earlier test data](https://github.com/googleapis/google-cloud-ruby/blob/master/google-cloud-storage/test/data/signer_v4_test_data.json#L69-L82) that I was using does not contain `tabs`.

quartzmo commented on google-cloud-storage/lib/google/cloud/storage/file/signer_v4.rb:
```diff
@@ -122,9 +122,10 @@ def issuer_and_signer issuer, client_email, signing_key, private_key
-                                       [k.downcase, v.strip.gsub(/\s+/, " ")]
-                                     end]
+            headers_arr = canonical_headers.map do |k, v|
+              [k.downcase, v.strip.gsub(/[^\S\t]+/, " ").gsub(/\t+/, "\t")]
```
@jskeet Does this whitespace replacement look ok to you? It passes the conformance tests.

jskeet replied:
Can't look now - will check tomorrow

jskeet replied:
Hmm... I'm not sure. In the C# code we collapse whitespace runs by replacing any run of whitespace with "the final character of that run"... which seems equally somewhat arbitrary, to be honest. (So "space tab" ends up as tab, but "tab space" ends up as space.)
I don't remember seeing details about this in the spec we were following - it may be worth following up on this. @frankyn Any thoughts?

frankyn replied:
Hi folks,
IIUC the problem breaks down to the following two questions:
Question 1: Which of the following is the correct result? Based on development the expected result is option 1.
1. `[tab][tab] -> [tab] && [space][space] -> [space]`
1. `[tab][tab] -> [space] && [space][space] -> [space]`

Question 2: Which of the following is the correct result? I'm not sure, but based on Question 1, I'd select option 2 to the expected result.
1. `[tab][tab][space][space] -> [tab][space]`
1. `[tab][tab][space][space] -> [space] && [space][space][tab][tab] -> [space]` 
1. `[tab][tab][space][space] -> [space] && [space][space][tab][tab] -> [tab]`

Background on handling whitespace wasn't well defined in the spec given the source material used: "Convert sequential spaces in the header value to a single space." is the documented expectation. 

If I'm on the same page as both of you (@jskeet, @quartzmo), PLMK and I'll follow-up with the GCS team.

jskeet replied:
From my implementation, I'd expect option 3 for your final set - but we really need to know what the server expects, and document that.

frankyn replied:
Thanks @jskeet. Sg, so I'm on the same page and will follow-up with the GCS team.

frankyn replied:
Internal bug: 136171758

quartzmo replied:
@jskeet Can we add the `[tab][tab][space][space]` scenario to the conformance test?

frankyn replied:
@quartzmo, as to not block HMAC SA support, can we move v4 signature issues into a separate PR?

jdpedrie replied:
@frankyn any word on this?

frankyn replied:
@jdpedrie, I pinged the GCS team last week but I haven't heard back yet. I would recommend not blocking support for HMAC SA for this question and splitting it off into a separate PR instead.

quartzmo commented:
@jkwlui I rebased this branch on master and fixed some small issues with the testing. I believe this PR is ready for final review/release. Do we still need to hold off from merging?  You will post here when the feature is off whitelist, correct?

frankyn commented:
@quartzmo the whitelist has been removed. Please merge and release when possible.

quartzmo commented:
@frankyn Got it, will do!

frankyn commented on google-cloud-storage/lib/google/cloud/storage/project.rb:
```diff
@@ -384,6 +385,73 @@ def create_bucket bucket_name, acl: nil, default_acl: nil,
+        #
+        # @return [Google::Cloud::Storage::HmacKey]
+        #
+        def create_hmac_key service_account_email, user_project: nil
```
@quartzmo could you make project_id an optional parameter? By default use project id instantiated at client creation.

quartzmo replied:
Yes, will do.

quartzmo replied:
Added `project_id` param to `Project#create_hmac_key`, `#hmac_key`, and `#hmac_keys`.

frankyn commented on google-cloud-storage/lib/google/cloud/storage/hmac_key.rb:
```diff
@@ -0,0 +1,312 @@
+        #
+        # @return [String]
+        #
+        def project_id
```
This should use the `project_id` returned from the `service.get_hmac_key` instead of `@gapi.project_id` otherwise the user provided `project_id` is lost.

quartzmo replied:
The `@gapi` instance variable here holds the `Google::Apis::StorageV1::HmacKeyMetadata` instance returned by the RPCs (including `service.get_hmac_key`), so that's OK isn't it?

frankyn replied:
oh. yes. Apologies @quartzmo, I completely misunderstood the gapi comment above. 

`@private The Google API Client object.` I thought it was a Google API client instance and not the HMACKeyMetadata instance.

frankyn commented on google-cloud-storage/lib/google/cloud/storage/hmac_key.rb:
```diff
@@ -0,0 +1,312 @@
+        attr_reader :secret
+
+        ##
+        # @private Creates a Notification object.
```
Nit: Use HmacKey instead of Notification object.

quartzmo replied:
Fixed.

frankyn commented on google-cloud-storage/lib/google/cloud/storage/hmac_key.rb:
```diff
@@ -0,0 +1,312 @@
+        #
+        def delete!
+          ensure_service!
+          @service.delete_hmac_key access_id,
```
set `project_id` parameter based on the value from `@project_id` when it's available in `HmacKey` class.

quartzmo replied:
Done, also in `#reload!`.

frankyn commented on google-cloud-storage/lib/google/cloud/storage/hmac_key.rb:
```diff
@@ -0,0 +1,312 @@
+          new.tap do |f|
+            f.gapi = gapi
+            f.service = service
+            f.user_project = user_project
```
Remember `project_id`

quartzmo replied:
The `f.gapi` attribute here is being assigned the `Google::Apis::StorageV1::HmacKeyMetadata` instance (including `project_id`) returned by the RPC, so that's OK isn't it?

frankyn commented on google-cloud-storage/lib/google/cloud/storage/hmac_key.rb:
```diff
@@ -0,0 +1,312 @@
+          ensure_service!
+          put_gapi = @gapi.dup
+          put_gapi.state = new_state
+          @gapi = service.update_hmac_key access_id, put_gapi,
```
Set `project_id` from `@project_id` when it's available.

quartzmo replied:
The `@gapi` instance variable here is being assigned the `Google::Apis::StorageV1::HmacKeyMetadata` instance (including `project_id` returned by the RPC, so that's OK isn't it?

quartzmo replied:
Done, also in `#reload!`.

frankyn commented on google-cloud-storage/lib/google/cloud/storage/hmac_key/list.rb:
```diff
@@ -0,0 +1,182 @@
+                             show_deleted_keys: nil, max: nil, user_project: nil
+            hmac_keys = new(Array(gapi_list.items).map do |gapi_object|
+              HmacKey.from_gapi_metadata gapi_object, service,
+                                         user_project: user_project
```
Set `project_id` value on HmacKey instance.

quartzmo replied:
The `gapi_object` variable here holds the `Google::Apis::StorageV1::HmacKeyMetadata` instance returned by the RPC and will be held in the `HmacKey` instance, so that's OK isn't it?

frankyn commented on google-cloud-storage/lib/google/cloud/storage/project.rb:
```diff
@@ -384,6 +385,91 @@ def create_bucket bucket_name, acl: nil, default_acl: nil,
+          gapi = service.create_hmac_key service_account_email,
+                                         project_id: project_id,
+                                         user_project: user_project
+          HmacKey.from_gapi gapi, service, user_project: user_project
```
Store `project_id` after create_hmac_key.

quartzmo replied:
The `gapi` variable here holds the `Google::Apis::StorageV1::HmacKeyMetadata` instance returned by the RPCs (including `service.get_hmac_key`) and will be held in the `HmacKey` instance, so that's OK isn't it?

frankyn commented on google-cloud-storage/lib/google/cloud/storage/project.rb:
```diff
@@ -384,6 +385,91 @@ def create_bucket bucket_name, acl: nil, default_acl: nil,
+        def hmac_key access_id, project_id: nil, user_project: nil
+          gapi = service.get_hmac_key \
+            access_id, project_id: project_id, user_project: user_project
+          HmacKey.from_gapi_metadata gapi, service, user_project: user_project
```
Store HmacKey after performing `get_hmac_key`

quartzmo replied:
Do you mean store the `HmacKey` instance in this `Project` instance? Generally these `Project` instances only hold connection/credentials state, not application state. (`Client` might have been a better name for these types.)

quartzmo replied:
@frankyn commented offline: "PR LGTM. Thanks Chris!"
